### PR TITLE
Reduce output buffer memory manager contention

### PR DESCRIPTION
**Pull request by pettyjamesm**
Addresses lock contention bottlenecks in `OutputBufferMemoryManager` by:
- Reducing the scope of the critical section in `#updateMemoryUsage(long)`
- Removing synchronization from `isBufferFull()` and `isOverutilized()`, these methods are already atomic reads
- Avoiding redundant notifications when the `blockedOnMemory` future does not change or is already completed (aka: not blocked)
- Avoiding spurious notifications from `setNoBlockOnFull()` when still blocked on memory
- Adds an unsynchronized fast path in `onMemoryAvailable()` to skip the lock acquisition entirely when the buffer is full

```
== RELEASE NOTES ==
General Changes
* Improve output buffer memory tracking to reduce lock contention 
```

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/execution/buffer/TestClientBuffer.java [hunk 1]
```diff
@@ -361,6 +361,48 @@ public void testReferenceCount()
...
+        ClientBuffer buffer = new ClientBuffer(TASK_INSTANCE_ID, BUFFER_ID, NOOP_RELEASE_LISTENER);
+        TestingPagesSupplier supplier = new TestingPagesSupplier();
+        supplier.addPage(createPage(0));
+        try {
+            ListenableFuture<BufferResult> result;
+            synchronized (buffer) {
+                result = buffer.getPages(0, sizeOfPages(1), Optional.of(supplier));
+            }
+            fail("Expected AssertionError to be thrown, are assertions enabled in your testing environment?");
+            assertTrue(getFuture(result, NO_WAIT).isEmpty(), "Code should not reach here");
```
Comment (aweisberg) [inline presto-main/src/test/java/com/facebook/presto/execution/buffer/TestClientBuffer.java:396]: I don't get why there should be something after fail?

Comment (pettyjamesm) [reply presto-main/src/test/java/com/facebook/presto/execution/buffer/TestClientBuffer.java:396]: Makes the reference used to satisfy some pedantic checkstyles but also the potential JIT DCE “optimizations”. Also, IDEs. Should be safe to omit, but I didn’t like how IntelliJ was judging me without it.

**Comment by aweisberg** [comment]
Makes sense to me. I think changing the assertions makes me wonder if we need to look harder at assertions in our release process. Seems like we want to run those checks before release.

**Review by aweisberg — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/execution/buffer/OutputBufferMemoryManager.java [hunk 2]
```diff
@@ -63,53 +66,79 @@ public OutputBufferMemoryManager(long maxBufferedBytes, Supplier<LocalMemoryCont
...
+            ListenableFuture<?> blockedOnMemory = systemMemoryContext.setBytes(currentBufferedBytes);
+            if (!blockedOnMemory.isDone()) {
+                if (this.blockedOnMemory != blockedOnMemory) {
+                    this.blockedOnMemory = blockedOnMemory;
+                    waitForMemory = blockedOnMemory; // only register a callback when blocked and the future is different
+                }
+            }
+            else {
+                this.blockedOnMemory = NOT_BLOCKED;
+                if (currentBufferedBytes <= maxBufferedBytes || !blockOnFull.get()) {
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/execution/buffer/OutputBufferMemoryManager.java:101]: why not use !isBufferFull()?

File: presto-main/src/main/java/com/facebook/presto/execution/buffer/OutputBufferMemoryManager.java [hunk 3]
```diff
@@ -63,53 +66,79 @@ public OutputBufferMemoryManager(long maxBufferedBytes, Supplier<LocalMemoryCont
...
+            if (closed) {
+                return;
+            }
+
+            currentBufferedBytes = bufferedBytes.addAndGet(bytesAdded);
+            ListenableFuture<?> blockedOnMemory = systemMemoryContext.setBytes(currentBufferedBytes);
+            if (!blockedOnMemory.isDone()) {
+                if (this.blockedOnMemory != blockedOnMemory) {
+                    this.blockedOnMemory = blockedOnMemory;
+                    waitForMemory = blockedOnMemory; // only register a callback when blocked and the future is different
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/execution/buffer/OutputBufferMemoryManager.java:96]: can you explain this logic?  What does it mean when the future is different?

**Review by rschlussel — COMMENTED** [review]
A couple questions on the main commit, but mostly looks good.  As discussed over slack let's remove the commits switching to use assertions since the performance benefit is minimal.

File: presto-main/src/main/java/com/facebook/presto/execution/buffer/OutputBufferMemoryManager.java (hunk 2 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/com/facebook/presto/execution/buffer/OutputBufferMemoryManager.java:101]: We already know the value of `currentBufferedBytes` and calling `!isBufferFull()` would be another volatile read from the `AtomicLong` inside of the critical section.

File: presto-main/src/main/java/com/facebook/presto/execution/buffer/OutputBufferMemoryManager.java (hunk 3 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/com/facebook/presto/execution/buffer/OutputBufferMemoryManager.java:96]: If the future is different it means that the memory reservation returned a new, not yet complete listener and we need to register a callback for when memory is available. If memory is available or it's not available but we've already registered a callback on the current instance- there's no need to register one. The previous implementation could attach arbitrarily many callbacks to the same future as well as callbacks to "not blocked" futures which dramatically increased contention on the output buffer itself.

Summary: only register a callback when we're waiting on memory and only do that one time per `ListenableFuture<?>` instance.

**Comment by pettyjamesm** [comment]
@rschlussel commits using `assert` have been dropped

**Review by rschlussel — APPROVED** [review]
Looks good, thanks!

**Pull request merged**